spring-cloud-context/src/main/java/org/springframework/cloud/bootstrap/BootstrapImportSelector.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2013-2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.bootstrap;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.springframework.context.EnvironmentAware;
+import org.springframework.context.annotation.DeferredImportSelector;
+import org.springframework.core.annotation.AnnotationAwareOrderComparator;
+import org.springframework.core.env.Environment;
+import org.springframework.core.io.support.SpringFactoriesLoader;
+import org.springframework.core.type.AnnotationMetadata;
+import org.springframework.util.ClassUtils;
+import org.springframework.util.StringUtils;
+
+/**
+ * This class uses {@link SpringFactoriesLoader} to load {@link BootstrapConfiguration}
+ * entries from {@code spring.factories}. The classes are then loaded so they can
+ * be sorted using {@link AnnotationAwareOrderComparator#sort(List)}.
+ * This class is a {@link DeferredImportSelector} so {@code @Conditional} annotations
+ * on imported classes are supported.
+ */
+public class BootstrapImportSelector implements EnvironmentAware, DeferredImportSelector {
+
+	private Environment environment;
+
+	@Override
+	public void setEnvironment(Environment environment) {
+		this.environment = environment;
+	}
+
+	@Override
+	public String[] selectImports(AnnotationMetadata metadata) {
+		ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
+		// Use names and ensure unique to protect against duplicates
+		List<String> names = new ArrayList<>(SpringFactoriesLoader
+				.loadFactoryNames(BootstrapConfiguration.class, classLoader));
+		names.addAll(Arrays.asList(StringUtils.commaDelimitedListToStringArray(
+				environment.getProperty("spring.cloud.bootstrap.sources", ""))));
+
+		List<Class<?>> sources = new ArrayList<>();
+		for (String name : names) {
+			Class<?> cls = ClassUtils.resolveClassName(name, null);
```

Comment:
Is class loading here still required? Seems as if Spring's `ConfigurationParser` considers ordering itself.

Reply:
Our ordering tests fail without it.

Reply:
@wilkinsona or @dsyer? I'm going to merge and we can revisit this particular nuance.

Reply:
I don't think the parser considers ordering of selected imports. Instead, the selector is expected to have done its own ordering. You could use the annotation metadata to get the `@Order` annotation and order things without class loading, but if this approach has worked up until now, it would appear to be good enough.

Reply:
The supplied annotation metadata looked like it was only for the import selector class. Is there a way to get the annotation metadata for a class name?

Reply:
> The supplied annotation metadata looked like it was only for the import selector class

Doh. Silly me. 

>  Is there a way to get the annotation metadata for a class name?

`org.springframework.core.type.classreading.SimpleMetadataReaderFactory.getMetadataReader(String)`

Reply:
This is what I came up with 1acb2b8b8bcaeac69284112b972215466190f03c